Kafka producer consumer setup

BeeMargarida opened the pull request:
| - | - |
| --- | --- |
| Issue | Related to https://github.com/ripe-tech/ripe-registry/issues/30 |
| Decisions | Initial setup for the repo, with API for kafka producer and consumer. The setup of the Kafka client and respective producer and consumer is made using `yonius.json`, and a consumer and producer is created. The consumer subscribes to the given topics, and the producer produces to the topic given when the produce request is made. <br><br> **Consumer**: <br> • Receives a callback function that will be called for each new message. It implements default onSuccess and onError logic, which can be overrided by functions `onSuccess` and `onError` in the `options`. <br> • The consumer is responsible for the retry logic by using a retry buffer, and will try the given number of retries until they expire. When that happens, the `sendError` method is called. br> • The consumer persists the failed messages while they still have retries. This persistence is made in the JSON file, which is updated when a new retry is made, a new message fails or a message expires their retries. <br> • To use the consumer: `bus.bind(<topic>, (message) => {...})` or `bus.bind(<topic>, { callback: (message) => {...}, ... })`<br><br> **Producer**: <br> • Sends an array of messages given, which will be converted to strings if they are provided as objects.  <br> • To use the producer: `bus.trigger(<topic>, <message>)`|
| Animated GIF | -- |

Commit:
Initial setup and consumer and producer Kafka entities

Commit:
Types

Commit:
Github templates

Commit:
FIx

joamag commented on js/consumer.js:
```diff
@@ -0,0 +1,151 @@
+import { conf } from "yonius";
+import * as fs from "fs";
+
+console.log(fs);
```
?

joamag commented on js/consumer.js:
```diff
@@ -0,0 +1,151 @@
+
+console.log(fs);
+
+export class Consumer {
```
Let's use the adapter pattern KafkaConsumer that respects the base consumer API

joamag commented on js/producer.js:
```diff
@@ -0,0 +1,48 @@
+import { conf } from "yonius";
+import { CompressionTypes } from "kafkajs";
+
+export class Producer {
```
Same here, `KafkaProducer` that respects a common `Producer` API

joamag commented:
@BeeMargarida on top of these changes an upper layer (much more abstract) for producing events should exist where the inner details (connection, etc.) are abstracted from the developer so that a producer node could just do something like

```javascript
bus.trigger("order.create", {order_number: "32132", ...})
```

all the singleton client-related logic is abstract by this API and the first time a trigger occurs a connection is established via the currently active engine and using the configuration details

joamag commented:
For the upper layer consumer API the equivalent: 

```javascript
bus.bind("order.created", (order) => {
});
```

Commit:
Kafka client singleton

Commit:
Producer and Consumer base classes

Commit:
Kafka adapters and base API

Commit:
Options object with success and error callbacks

Commit:
Subscribe to multiple topics

Commit:
Default onSuccess and onError consumer functions

Commit:
Fixes

Commit:
try catch for callback

Commit:
Fix types

Commit:
Refactor

Commit:
Initialization and usage in README

BeeMargarida commented:
@joamag I implemented a default `onSuccess` and `onError` logic, which can be overwritten by the options object in the `bind` and `trigger` methods. Regarding errors in the callback of bind, I changed the implementation to a try catch.

I have a doubt however, regarding the parameters for the Kafka client, consumer and producer. For now they are using the values in the `yonius.json`, but would it be better to allow them to be overridden by a `kwargs` in the API constructor?

Commit:
Fix package

Commit:
Support for not array in producer

Commit:
Try/catch on retry

Commit:
Retry persistence dir from configs

Commit:
Fix

Commit:
Multiple callbacks for different topics, support for dynamic subscription

joamag commented on README.md:
````diff
@@ -1,3 +1,20 @@
+## Usage
+
+```javascript
+bus.bind("order", (message) => { ... })
````
separate bind from trigger and make the second parameter either a function or an Object

joamag commented on js/kafkaClient.js:
```diff
@@ -0,0 +1,27 @@
+import { Kafka } from "kafkajs";
+
+export const KafkaClient = (function() {
+    var client = null;
```
var ?

joamag commented on js/kafkaConsumer.js:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        if (!fs.existsSync(`${dir}/retry.json`)) return;
+        const data = fs.readFileSync(`${dir}/retry.json`, { encoding: "utf-8" });
```
use the async fs api

joamag replied:
https://nodejs.org/api/fs.html#fs_fs_promises_api

joamag commented on js/kafkaConsumer.js:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    async _retry(callback, options = {}) {
+        if (this.retryBuffer.length === 0) return;
```
encapsulate this around a `KafkaRetryConsumer` abstraction that inherits from `KafkaConsumer`

joamag commented on js/kafkaProducer.js:
```diff
@@ -0,0 +1,54 @@
+
+        await this.producer.send({
+            topic: topic,
+            acks: conf("KAFKA_PRODUCER_ACKS", 0),
```
try to make this as part of the constructor logic

joamag replied:
`acks: options.acks === undefined ? this.acks : options.acks`

joamag commented on js/kafkaProducer.js:
```diff
@@ -0,0 +1,54 @@
+    constructor() {
+        super();
+
+        const kafkaClient = KafkaClient.getInstance();
```
`options.acks === undefined ? conf("KAFKA_PRODUCER_ACKS", trure) : options.acks`

Commit:
More specific configuration of creation

Commit:
Removed retry logic from consumer

Commit:
Options for bus configuration in constructors

Commit:
Kafka retry consumer and producer

Commit:
Async fs calls

Commit:
Fix

Commit:
Comments

Commit:
Refactor and documentation

Commit:
Refactor

joamag committed:
Small fixes on examples

joamag committed:
Removed scpaces between pipes

joamag committed:
Code refactor for configuration

joamag committed:
New file naming

Commit:
Refactor configuration variables

Commit:
Using owner in consumer triggers

Commit:
Unit tests

Commit:
Fix

Commit:
Fix in retry dir

Commit:
Refactor

Commit:
Pretty

joamag commented:
👍
